src/EnergyPlus/WaterThermalTanks.cc

```diff
@@ -1357,69 +1319,51 @@ namespace WaterThermalTanks {
-                            ErrorsFound = true;
-                        }
+                        WaterHeaterDesuperheater(DesuperheaterNum).ReclaimHeatingSourceIndexNum = VariableSpeedCoils::GetCoilIndexVariableSpeed(cAlphaArgs(9), cAlphaArgs(10), errFlag);
+                        if (allocated(HeatReclaimDXCoil)) ValidSourceType(DesuperheaterNum) = true;
+                    } else if (UtilityRoutines::SameString(cAlphaArgs(9), "Coil:Cooling:WaterToAirHeatPump:EquationFit")) {
+                        WaterHeaterDesuperheater(DesuperheaterNum).ReclaimHeatingSource = COIL_AIR_WATER_HEATPUMP_EQ;
+                        WaterHeaterDesuperheater(DesuperheaterNum).ReclaimHeatingSourceIndexNum = WaterToAirHeatPumpSimple::GetCoilIndex(cAlphaArgs(9), cAlphaArgs(10), errFlag);
+                        if (allocated(DataHeatBalance::HeatReclaimSimple_WAHPCoil)) ValidSourceType(DesuperheaterNum) = true;
                     } else {
+                        ShowSevereError(cCurrentModuleObject + " = " + WaterHeaterDesuperheater(DesuperheaterNum).Name + ':');
+                        ShowContinueError(" desuperheater can only be used with Coil:Cooling:DX:SingleSpeed, ");
+                        ShowContinueError(
+                            " Coil:Cooling:DX:TwoSpeed, Coil:Cooling:DX:TwoStageWithHumidityControlMode, Coil:Cooling:DX:VariableSpeed, "
+                            "Coil:Cooling:WaterToAirHeatPump:EquationFit, Refrigeration:CompressorRack,");
+                        ShowContinueError(" Refrigeration:Condenser:AirCooled ,Refrigeration:Condenser:EvaporativeCooled, ");
+                        ShowContinueError(" or Refrigeration:Condenser:WaterCooled.");
+                        ShowContinueError(" Invalid desuperheater heat source object: " + cAlphaArgs(9) + " \"" + cAlphaArgs(10) + "\"");
+                        ErrorsFound = true;
+                    }
+                    if (errFlag) {
+                        ShowContinueError("...occurs in " + cCurrentModuleObject + '=' + WaterHeaterDesuperheater(DesuperheaterNum).Name);
+                        ErrorsFound = true;
+                    }
+
+                    if (WaterHeaterDesuperheater(DesuperheaterNum).ReclaimHeatingSourceIndexNum == 0) {
                         ShowSevereError(cCurrentModuleObject + ", \"" + WaterHeaterDesuperheater(DesuperheaterNum).Name +
-                                        "\" invalid desuperheater heat source object: " + cAlphaArgs(9) + " \"" + cAlphaArgs(10) + "\"");
+                                        "\" desuperheater heat source object not found: " + cAlphaArgs(9) + " \"" + cAlphaArgs(10) + "\"");
                         ErrorsFound = true;
                     }
 
+                    WaterHeaterDesuperheater(DesuperheaterNum).HeatingSourceType = cAlphaArgs(9);
+                    WaterHeaterDesuperheater(DesuperheaterNum).HeatingSourceName = cAlphaArgs(10);
```

Comment:
These two lines need to move up before the long if blocks (after line 1303), because `WaterHeaterDesuperheater(DesuperheaterNum).HeatingSourceName` is used in the index lookups. I tested this locally with CoilWaterDesuperheating.idf and this fixes the error.

Reply:
Ok I will look into this issue. But I think these two parts are not doing the same function, so I would retake the combination. What do you think?

Reply:
In the old code, the first part checks to see if the source object exists, and sets these two variables:
```
WaterHeaterDesuperheater(DesuperheaterNum).HeatingSourceType = cAlphaArgs(9);
WaterHeaterDesuperheater(DesuperheaterNum).HeatingSourceName = cAlphaArgs(10);
```
For the dxcoils, the use of functions like `GetDXCoilIndex` and `VariableSpeedCoils::GetCoilIndexVariableSpeed` will force the coil inputs to be processed if they haven't been yet. But the old code does not save the returned index value here. Instead, it loops through all of the coils again in the second part trying to match up the name. And then it saves the index in `WaterHeaterDesuperheater(DesuperheaterNum).ReclaimHeatingSourceIndexNum`.

You an confirm this by stepping through the old code in the debugger and watch the index value returned in the first section and see that it's the same value found in the second section.  Your new code merges the two parts well, except that `HeatingSourceType` and `HeatingSourceName` need to be set before the getcoilindex calls.

Reply:
I got what you're saying. You're correct, the ``HeatReclaimDXCoil`` was saving all of the available coils not only the ''heat reclaim'' ones, I was misled by its name, so they're doing the same thing you're right.